include/tscore/LogThrottle.h

```diff
@@ -0,0 +1,113 @@
+/** @file
+
+  LogThrottle declaration.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include "ThrottledMessage.h"
+
+#include <cstdint>
+#include <unordered_map>
+
+/** Provides an interface for throttling messages at a certain rate.
+ *
+ * Note that each id represents its own throttling rate. This allows the
+ * flexibility of throttling a set of messages together. For instance:
+ *
+ * LogThrottle::status(1, "This is throttled.");
+ *
+ * LogThrottle::status(2, "This is throttled independantly from id 1 above.");
+ *
+ * LogThrottle::note(1, "This, a different message at a different log level,
+ *     "is throttled with the above status with id 1.");
+ */
+class LogThrottle
+{
+public:
+  /** Log a throttled Status message.
+   *
+   * @param[in] id The identifier for the log message being throttled.
+   */
+  template <class... Args>
+  static void
+  status(uint32_t id, Args... args)
+  {
+    auto &message = get_or_create_message(id);
+    message.ThrottleStatus(args...);
```

Comment:
Hmm, why aren't you using perfect forwarding with these?

Reply:
Heh. I dinged on that too :-P